source/develop/release-management/features/sla/hosted-engine-agent-offloading.html.md

```diff
@@ -0,0 +1,53 @@
+## Owner
+
+*   Name: [Denis Chaplygin](https://github.com/akashihi)
+*   Email: <dchaplyg@redhat.com>
+
+## Detailed Description
+
+The specific goals associated with this feature are as follows:
+
+*   Move storage mounting/unmounting from agent to the broker.
+*   Broker should start/stop domain monitor on agent's command. Domain monitor could be implemented as submonitor.
+    
+    Domain monitoring needs to be changed drastically. When storage validation approach was changed (See [Documentation/External references](develop/release-management/features/sla/hosted-engine-agent-offloading/#documentation)), domain monitoring become an important for domain to be valied. Non-monitored domains are not reported as valid anymore. 
+    Therefore, old approach, when domains monitoring was active only outside of the LocalMaintenance mode is not applicable anymore. I propose it to be changed in the following way - domain monitoring still should be disabled in the LocalMaintenance mode. While domain monitoring is disabled, actual domain validation will be skipped and domain will be reported as valid. Broker will still try to update status (and read statues of the other cluster participants) even in the maintenance mode, so it will have a chance to block on broken storage. Thus, all I/O should be moved to a separate thread, so broker will not block on I/O issues, only that thread.
+*   Move sanlock acquire/release activities form agent to the broker.
+*   As steps above introduce new possible 'half-configured' state, FSM should be extended with that state.
+*   Agent should post state to the broker only and broker must publish it on storage if possible.
+
+    Broker will keep current status in memory and reply to agent, hosted-engine setup and vdsm using that in-memory state. Agent will call broker's RPC to update that state. Upon in-memory state update separate disk writing thread will be notified. That separate disk writing thread will be started during broker start-up sequence and it's only purpose is to write the state to the storage and block in case of storage failures, while letting other broker code to run. I plan to use deque with size one to pass the current state to disk writing thread and use Conditions to synchronize threads.
+*   OVF extraction should be moved from agent to the broker.
+	At the moment OVF are extractred on every short monitoring loop run by calling `_initialize_storage_images` that call `refresh_vm_conf`, which actually updates `vm.conf` from the OVF. I plan to wrap `refresh_vm_conf` function to the submonitor and specify `vm.conf` update period as submonitor parameter, so agent will be able to control it. Submonitor will update the `vm.conf` peridically and report time passed from the last update. We can use that time in score calculation in the futue.
```

Comment:
- "At the moment OVF are extractred" -> "At the moment, OVF are extracted"
- "that call `refresh_vm_conf`" -> ", which calls `refresh_vm_conf`"
- "I plan to wrap `refresh_vm_conf` function to the submonitor and specify `vm.conf` update period as submonitor parameter, so agent will be able to control it." -> "I plan to move the `refresh_vm_conf` function to the submonitor and specify the `vm.conf` update period as a submonitor parameter, so the agent will be able to control it."
- "Submonitor" -> "The submonitor"
- "peridically and report time passed from the last update. We can use that time in score calculation in the futue." -> "periodically and report the time elapsed since the last update, which can then be used in score calculations in the future."

Reply:
Done